### PR TITLE
Feature/basic query (& query-opt extension)

### PR DESCRIPTION
**Pull request by bplatz**
This PR extends the work that was done for query-optimizations here: https://github.com/fluree/db/pull/164

It expands the types of queries that qualify for :simple-subject-crawl optimization, fixes a few bugs in the original PR, and now does enough it can completely replace the query execution engine we had specifically for 'basic query'. Now all basic queries are transpiled into analytical queries and processed by the new engine (as they all qualify for :simple-subject-crawl optimizations).

It also introduce multiple :vars bindings, but today only works for :simple-subject-crawl style queries, where a single var can have a vector value of possible solutions (e.g {select: {?s [*]} where: [[?s _id ?ids]] vars: {?ids [[_user/username "someuser"], 456456456, 345345435]}}

### CODE REVIEW

File: src/fluree/db/query/subject_crawl/rdf_type.cljc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                      (query-range/intersects-range? node range-set))
+        query-xf    (where-subj-xf {:start-test  >=
+                                    :start-flake fflake
+                                    :end-test    <=
+                                    :end-flake   lflake
+                                    :return-type :flake-by-sid})
+        resolver    (index/->CachedTRangeResolver conn (:spot novelty) t t (:async-cache conn))
+        tree-chan   (index/tree-chan resolver spot in-range? query-range/resolved-leaf? 1 query-xf error-ch)
+        return-chan (async/chan 10 (partition-by flake/s))]
+    (async/go-loop []
```
Comment (zonotope) [inline src/fluree/db/query/subject_crawl/rdf_type.cljc:61]: I think this go-loop can be:

```clojure
(async/pipeline 10 return-chan cat tree-chan)
```

File: src/fluree/db/query/subject_crawl/subject.cljc [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+                                    :end-flake   lflake
+                                    ;; if looking for pred + obj, but pred is not indexed, then need to use :psot and filter for 'o' values
+                                    :xf          (when filter-fn
+                                                   (map (fn [flakes]
+                                                          (filter filter-fn flakes))))})
+        resolver    (index/->CachedTRangeResolver conn (get novelty idx) t t (:async-cache conn))
+        tree-chan   (index/tree-chan resolver idx-root in-range? query-range/resolved-leaf? 1 query-xf error-ch)
+        return-chan (async/chan 10 (comp (map flake/s)
+                                         (dedupe)))]
+    (async/go-loop []
```
Comment (zonotope) [inline src/fluree/db/query/subject_crawl/subject.cljc:52]: This could be a pipeline too

**Review by zonotope — APPROVED** [review]
LGTM 👍🏾  besides a few minor suggestions.

File: src/fluree/db/query/analytical_filter.cljc [hunk 3]
```diff
@@ -91,37 +91,47 @@
...
+                                                 (throw (ex-info (str "Invalid symbol: " arg
+                                                                      " used in function argument: " (pr-str func))
+                                                                 {:status 400 :error :db/invalid-fn}))
+                                                 [(conj args* arg) (conj vars arg)])
+                                 (or (string? arg)
+                                     (number? arg)
+                                     (boolean? arg)
+                                     (nil? arg)
+                                     #?(:clj  (= (type arg) java.util.regex.Pattern)
+                                        :cljs (regexp? arg))) [(conj args* arg) vars]
```
[the remark this answers is not in the document]
Comment (bplatz) [reply src/fluree/db/query/analytical_filter.cljc:128]: Sure will. FWIW This namespace is all code from the split older namespace, I didn't really review this code except to move it.

File: src/fluree/db/query/subject_crawl/rdf_type.cljc (hunk 1 above)
Comment (bplatz) [reply src/fluree/db/query/subject_crawl/rdf_type.cljc:61]: I'm not sure... here I'm doing a buffer of 10 vs a pipeline would be parallel processing of 10. I'm going to leave as is for now but we can discuss if there is a better way. The buffer is intentional to slightly 'read ahead' of the data needed downstream as it should be fairly cheap to do so and then ultimately pre-load the next index before it is needed.

**Pull request merged**